File: 算法/3.KMP算法.cpp
#include<iostream>

using namespace std;

void Getnext(string str , int next[])
{
    int j , k;
    j = 0;
    k = -1;
    next[j] = k;
    while (j < str.size()-1)
    {
        if (k == -1 || str[k] == str[j])
        {
            k++;j++;
            next[j] = k;
        }
        else
        {
            k = next[k];
        }
    }
}


int Kmp(string &str , string &ex_str)
{
    int m = str.size();
    int n = ex_str.size();
    int next[ex_str.size()];
    Getnext(ex_str , next);
    int i = 0 ; int j = 0;
    while (i < m && j < n)
    {
        if (j == -1 || str[i] == ex_str[j])
        {
            i++;
            j++;
        }
        else
        {
            j = next[j];
        }
    }
    if (j >= n)
    {
        return (i - n);
    }
    else
    {
        return -1;
    }
}



int main()
{
    string str = "abcabeabcabcmn";
    string temp = "abcabcmn";
    int pos;
    pos = Kmp(str , temp);
    cout << pos << endl;
    return 0;
}

File: 算法/3.KMP算法.py
def GetNext(str):
    next = [0] * len(str)
    k = -1
    i = 0
    next[i] = k
    while i < len(str)-1:
        if k == -1 or str[k] == str[i]:
            k += 1
            i += 1
            next[i] = k
        else:
            k = next[k]
    return next

def KMP( father_str , son_str):
    i = 0
    j = 0
    next = GetNext(son_str)
    while i < len(father_str) and j < len(son_str):
        if j == -1 or father_str[i] == son_str[j]:
            i += 1
            j += 1
        else:
            j = next[j]
    if j >= len(son_str):
        return i - len(son_str)
    else:
        return -1
    

s1 = "abcabcdefmnabcdmn"
s2 = "abcdmn"
print(KMP(s1 , s2))